Ticket4918 sans dls

Alistair-McGann-Tessella opened the pull request:
## Description of work:
Adds a PCASpy wrapper around the LSI vendor software

## Ticket:
https://github.com/ISISComputingGroup/IBEX/issues/4918

## Acceptance criteria:
- [ ] LSI software parameters are available as PVs
- [ ] User can start/stop the correlator
- [ ] Correlation function, time lags, trace A and B and metadata saved to a .txt file
     - [ ] These values are made available for use in #4999
- [ ] Device reads disconnected error when the correlator is not connected
- [ ] Driver structure is adequately documented on the wiki

Alistair-McGann-Tessella committed:
Add driver code

Alistair-McGann-Tessella committed:
Ignore more files

Alistair-McGann-Tessella committed:
IOC can serve a PV value

Alistair-McGann-Tessella committed:
Add print to show IOC has started

Alistair-McGann-Tessella committed:
PV sets are now threaded

Alistair-McGann-Tessella committed:
Add call to LSI driver code when setting

Alistair-McGann-Tessella committed:
Make integer fields rounded floats to stop errors

Alistair-McGann-Tessella committed:
Change check for integer pvs

Alistair-McGann-Tessella committed:
Add error message pv

Alistair-McGann-Tessella committed:
Fix spelling error

Alistair-McGann-Tessella committed:
Add read fore correlation type

Alistair-McGann-Tessella committed:
Add ability to write to different pv

Alistair-McGann-Tessella committed:
Add new data structure to keep track of PV attributes

Alistair-McGann-Tessella committed:
Clean up and return correct enum values

Alistair-McGann-Tessella committed:
Do check on the output value rather than the PV

Alistair-McGann-Tessella committed:
Tidying up

Alistair-McGann-Tessella committed:
Add enum lookup

Alistair-McGann-Tessella committed:
Add read function lookup

Alistair-McGann-Tessella committed:
Move PVConfig to new file

Alistair-McGann-Tessella committed:
Remove duplicated code

Alistair-McGann-Tessella committed:
Tidy up

Alistair-McGann-Tessella committed:
Add filepath PV

Alistair-McGann-Tessella committed:
Add take data method

Alistair-McGann-Tessella committed:
Can save the collected data

Alistair-McGann-Tessella committed:
Add is connected PV

Alistair-McGann-Tessella committed:
Update the RUNNING pv when device is taking data

Alistair-McGann-Tessella committed:
Add metadata pvs and tidy up

Alistair-McGann-Tessella committed:
Many changes; add customisable filename

Alistair-McGann-Tessella committed:
Add trace channels

Alistair-McGann-Tessella committed:
Save raw traces and metadata

Alistair-McGann-Tessella committed:
Add egu fields

Alistair-McGann-Tessella committed:
Add new database for EGU fields

Alistair-McGann-Tessella committed:
Add array fields

Alistair-McGann-Tessella committed:
Add function to update an array pv

Alistair-McGann-Tessella committed:
Add val, sevr and stat fields

Alistair-McGann-Tessella committed:
Remove unused code

Alistair-McGann-Tessella committed:
Update errormsg pv more often

Alistair-McGann-Tessella committed:
Add sim and disable PVs

Alistair-McGann-Tessella committed:
Add setpoint pvs

Alistair-McGann-Tessella committed:
Most classes work properly after refactor

Alistair-McGann-Tessella committed:
Move writing the new value to the device

Alistair-McGann-Tessella committed:
PV read write refactor complete?

Alistair-McGann-Tessella committed:
Re add logging

Alistair-McGann-Tessella committed:
Remove unused function

Alistair-McGann-Tessella committed:
Add start/stop as bool pvs

Alistair-McGann-Tessella committed:
Only need to call set_instrument once

Alistair-McGann-Tessella committed:
Remove unused code

Alistair-McGann-Tessella committed:
Stop button sends correctly

Alistair-McGann-Tessella committed:
Add LSI to archiver

Alistair-McGann-Tessella committed:
Can change IOC number on boot

Alistair-McGann-Tessella committed:
VAL fields now update properly

Alistair-McGann-Tessella committed:
Add high interest to PVs

Alistair-McGann-Tessella committed:
device reads disconnected on no data

Alistair-McGann-Tessella committed:
Need to convert number of repetitions to integer

Alistair-McGann-Tessella committed:
Remove unused function

Alistair-McGann-Tessella committed:
Hard code file path

Alistair-McGann-Tessella committed:
Read if device is connected on boot

Alistair-McGann-Tessella committed:
Add timestamp instead of repetition

Alistair-McGann-Tessella committed:
The polling for device connection does not work

Alistair-McGann-Tessella committed:
Move data wrangling to its own function

Alistair-McGann-Tessella committed:
Add disconnected alarms

Alistair-McGann-Tessella committed:
Update param values with sanitised values

Alistair-McGann-Tessella committed:
Add makefile

Alistair-McGann-Tessella committed:
Remove vendor python script

DominicOram commented on LSi_Correlator.py:
```diff
@@ -0,0 +1,425 @@
+from datetime import datetime
+
+sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), os.pardir)))
+sys.path.insert(1, 'C:\\Instrument\\Dev\\LSI-Correlator')
```
As discussed, we need to think about where the private repository goes on the instrument. Two solutions that come to mind are:
* The code in this repo gets put into the private one and we just use the private one only
* The private repo is a submodule of this repo

DominicOram commented on LSi_Correlator.py:
```diff
@@ -0,0 +1,425 @@
+
+sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), os.pardir)))
+sys.path.insert(1, 'C:\\Instrument\\Dev\\LSI-Correlator')
+sys.path.insert(2, 'C:\\Instrument\\Apps\\EPICS\\ISIS\\inst_servers\\master\\')
```
Hard coded path, you should be able to use the `EPICS_KIT_ROOT` environment variable.

DominicOram commented on pvdb.py:
```diff
@@ -0,0 +1,112 @@
+sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), os.pardir)))
+from server_common.utilities import char_waveform
+
+sys.path.insert(1, 'C:\\Instrument\\Dev\\LSI-Correlator')
```
Same as above, need to find a better home for this.

DominicOram commented on pvdb.py:
```diff
@@ -0,0 +1,112 @@
+
+
+PARAM_FIELDS_BINARY = {'type': 'enum', 'enums': ["NO", "YES"]}
+PARAM_FIELDS_BINARY = {'type': 'enum', 'enums': ["NO", "YES"], 'info_field': {'archive': 'VAL', 'INTEREST': 'HIGH'}}
```
`PARAM_FIELDS_BINARY` is being defined twice

DominicOram commented on pvdb.py:
```diff
@@ -0,0 +1,112 @@
+from pcaspy.alarm import AlarmStrings, SeverityStrings
+
+sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), os.pardir)))
+from server_common.utilities import char_waveform
```
`char_waveform` isn't being used

DominicOram commented on pvdb.py:
```diff
@@ -0,0 +1,112 @@
+    PvNames.SAMPLE_TEMP: {'type': 'float', 'unit': 'C', 'info_field': {'archive': 'VAL', 'INTEREST': 'HIGH'}},
+    PvNames.SOLVENT_VISCOSITY: {'type': 'float', 'unit': 'mPas', 'info_field': {'archive': 'VAL', 'INTEREST': 'HIGH'}},
+    PvNames.SOLVENT_REFRACTIVE_INDEX: {'type': 'float', 'unit': '', 'info_field': {'archive': '', 'INTEREST': 'HIGH'}},
+    PvNames.LASER_WAVELENGTH: {'type': 'float', 'unit': 'nm', 'info_field': {'archive': 'VAL', 'INTEREST': 'HIGH'}},
```
These are all basically the same but with different units, worth pulling out into a variable?

DominicOram commented on pvdb.py:
```diff
@@ -0,0 +1,112 @@
+ALARM_SEVR_PV_FIELDS = {'type': 'enum', 'enums': SeverityStrings}
+
+
+class PvNames(object):
```
This should be an ENUM (https://docs.python.org/3.7/library/enum.html)

DominicOram commented on pvdb.py:
```diff
@@ -0,0 +1,112 @@
+    return field_database
+
+
+STATIC_PV_DATABASE.update(add_setpoint_pvs(STATIC_PV_DATABASE))
```
Does everything in `STATIC_PV_DATABASE` need a `SP` equivalent?

DominicOram commented on pvdb.py:
```diff
@@ -0,0 +1,112 @@
+        if 'unit' in pv_definition:
+            field_database.update({"{PV_name}.EGU".format(PV_name=pv_name): {'type': 'string', 'value': pv_definition['unit']}})
+
+    return field_database
```
This seems useful to anything that uses PCASpy, can we put it centrally or better yet get it into pcaspy?

DominicOram commented on LSi_Correlator.py:
```diff
@@ -0,0 +1,425 @@
+    # of how it achieves this.
+    ip_address = '127.0.0.1'
+    firmware_revision = '4.0.0.3'
+    filepath = ""
```
Can you change this to have a sensible default (maybe the folder this file is in?) until the macro ticket is done? Just so that it will run without needing changes.

DominicOram commented on LSi_Correlator.py:
```diff
@@ -0,0 +1,425 @@
+    A driver for the LSi Correlator
+    """
+
+    def __init__(self, pv_prefix, host: str, firmware_revision: str, filepath: str):
```
`pv_prefix` is unused

DominicOram commented on LSi_Correlator.py:
```diff
@@ -0,0 +1,425 @@
+
+        self.SettingPVs = get_pv_configs(self.device)
+
+        self.PVValues = {
```
These variables should be snake_case

DominicOram replied:
There are effectively 3 different dictionaries in this object, all very similar:
* SettingsPVs
* PVValues
* pvDB (the one you use through `setParam` and `getParam`)

I think `SettingsPVs` probably needs a better name. I'm not sure I understand the difference between `PVValues` and `pvDB`

DominicOram commented on LSi_Correlator.py:
```diff
@@ -0,0 +1,425 @@
+
+        for pv, preset in self.PVValues.items():
+            # Write defaults to device
+            print("setting {} to {}".format(pv, preset))
```
This should be `print_and_log`

DominicOram commented on LSi_Correlator.py:
```diff
@@ -0,0 +1,425 @@
+            status = Alarm.TIMEOUT_ALARM
+        else:
+            severity = Severity.INVALID_ALARM
+            status = Alarm.TIMEOUT_ALARM
```
This if statement doesn't do anything?

DominicOram commented on LSi_Correlator.py:
```diff
@@ -0,0 +1,425 @@
+
+        if reason in self.SettingPVs:
+            # Need to convert internal state to PV (e.g. enum number)
+            pv_value = self.SettingPVs[reason].convert_to_pv(self.PVValues[reason])
```
We're calling the convert here in the read and when we do the write in `update_pv_handler`. Does that mean they're going through the converter twice?

DominicOram commented on LSi_Correlator.py:
```diff
@@ -0,0 +1,425 @@
+        trace_A = np.asarray(self.device.TraceChA)
+        trace_B = np.asarray(self.device.TraceChB)
+
+        return Corr, Lags, trace_A, trace_B
```
snake_case please

DominicOram reviewed (changes requested):
Generally works but could do with a bit of tidying up, see comments.

DominicOram commented:
Also, as discussed sending a hex value to a number of the PVs will be converted to floats. e.g. `caput %MYPVPREFIX%:LSICORR_01:SCATTERING_ANGLE:SP b`. Can we make sure the user cant do this in the OPI at least and see if you can fix in the server if easy?

Alistair-McGann-Tessella committed:
Tidy up import

Alistair-McGann-Tessella committed:
Tidy up pv definitions

Alistair-McGann-Tessella committed:
Can unset a disconnected alarm

Alistair-McGann-Tessella committed:
Add some snake case and default dir

Alistair-McGann-Tessella committed:
Static pv database now builds

Alistair-McGann-Tessella committed:
IOC boots with new class

Alistair-McGann-Tessella committed:
Setpoints now work correctly

Alistair-McGann-Tessella committed:
Remove commented code, tidy up

Alistair-McGann-Tessella committed:
Update NORD field directly in the paramter

Alistair-McGann-Tessella committed:
Starting to get rid of pvvalues

Alistair-McGann-Tessella committed:
Default values dont reference pvvalues dict

Alistair-McGann-Tessella committed:
Rename function and fix typo

Alistair-McGann-Tessella committed:
Update start pv after runs finished

Alistair-McGann-Tessella committed:
Remove PVConfig file

Alistair-McGann-Tessella committed:
Remove unused code

Alistair-McGann-Tessella committed:
Rename class, remove unused code

Alistair-McGann-Tessella committed:
Docstrings

Alistair-McGann-Tessella committed:
Refactor pcaspy code out of ibex code

Alistair-McGann-Tessella committed:
Add helper function

Alistair-McGann-Tessella committed:
Remove unused argument

Alistair-McGann-Tessella committed:
Errormsg seems to update correctly now

Alistair-McGann-Tessella committed:
Need to go through both input sanitisers

Alistair-McGann-Tessella committed:
Remove accidentally commited file

Alistair-McGann-Tessella committed:
Add pycache to git ignore file

Alistair-McGann-Tessella committed:
Point lsi vendor code to proper submodule

Alistair-McGann-Tessella committed:
Tidy up comments

Alistair-McGann-Tessella committed:
FILENAME has a setpoint pv

DominicOram committed:
Removed unused code

DominicOram reviewed (approved):
Looks much better, thanks. Only question is what is the difference between `get_pv_value` and `read`? They seem to be very similar?

Alistair-McGann-Tessella committed:
Clean up function

Alistair-McGann-Tessella committed:
Merge branch 'Ticket4918_sans_dls' of https://github.com/ISISComputingGroup/EPICS-LSICorrelator into Ticket4918_sans_dls

Alistair-McGann-Tessella commented:
> 
> 
> Looks much better, thanks. Only question is what is the difference between `get_pv_value` and `read`? They seem to be very similar?

They are quite similar, I've changed `get_pv_value` and renamed it to hopefully give the two some distinction
